File: Src/Icm.ContextConsole/Application/StandardApplication.vb
Imports Icm.Localization
Imports Icm.Collections
Imports Icm.Tree

Public Class StandardApplication
    Implements IApplication
    Private ReadOnly _internalLocRepo As ILocalizationRepository
    Private ReadOnly _externalLocRepo As ILocalizationRepository
    Private ReadOnly _interactor As IInteractor
    Private _currentContextNode As ITreeNode(Of IContext)
    Private _rootContextNode As ITreeNode(Of IContext)
    Private _contextsCache As IEnumerable(Of IContext)
    Private ReadOnly _tokenParser As ITokenParser

    Public Sub New(rootContextType As Type,
                  Optional tokenParser As ITokenParser = Nothing,
                  Optional interactor As IInteractor = Nothing,
                  Optional intLocRepo As ILocalizationRepository = Nothing,
                  Optional extLocRepo As ILocalizationRepository = Nothing
                  )
        MyClass.New(
            New ReflectionContextTreeBuilder(rootContextType),
            tokenParser,
            interactor,
            intLocRepo,
            extLocRepo
        )
    End Sub

    Public Sub New(rootContextNode As TreeNode(Of Type),
                  Optional tokenParser As ITokenParser = Nothing,
                  Optional interactor As IInteractor = Nothing,
                  Optional intLocRepo As ILocalizationRepository = Nothing,
                  Optional extLocRepo As ILocalizationRepository = Nothing
                  )
        MyClass.New(
            New TypeContextTreeBuilder(rootContextNode),
            tokenParser,
            interactor,
            intLocRepo,
            extLocRepo
        )
    End Sub

    Public Sub New(
                  treeBuilder As IContextTreeBuilder,
                  Optional tokenParser As ITokenParser = Nothing,
                  Optional interactor As IInteractor = Nothing,
                  Optional intLocRepo As ILocalizationRepository = Nothing,
                  Optional extLocRepo As ILocalizationRepository = Nothing
                  )

        If tokenParser Is Nothing Then
            _tokenParser = New StandardTokenParser
        Else
            _tokenParser = tokenParser
        End If
        If intLocRepo Is Nothing Then
            intLocRepo = New ResourceLocalizationRepository(My.Resources.ResourceManager)
        End If
        If extLocRepo Is Nothing Then
            extLocRepo = New DictionaryLocalizationRepository()
        End If
        If interactor Is Nothing Then
            interactor = New StreamsInteractor(intLocRepo)
        End If
        If tokenParser Is Nothing Then
            tokenParser = New StandardTokenParser()
        End If
        treeBuilder.Application = Me
        _interactor = interactor
        _internalLocRepo = intLocRepo
        _externalLocRepo = extLocRepo
        ApplicationPrompt = System.Reflection.Assembly.GetExecutingAssembly.CodeBase
        DependencyResolver = New StandardDependencyResolver
        ' Stablish root context in the last place so that the IContext.Initialize routine
        ' can access the former application values.
        SetRootContextNode(treeBuilder.GetTree)
    End Sub

    Property ApplicationPrompt As String Implements IApplication.ApplicationPrompt

    Property DependencyResolver As IDependencyResolver Implements IApplication.DependencyResolver

    ReadOnly Property ExternalLocRepo() As ILocalizationRepository Implements IApplication.ExternalLocRepo
        Get
            Return _externalLocRepo
        End Get
    End Property

    ReadOnly Property InternalLocRepo() As ILocalizationRepository Implements IApplication.InternalLocRepo
        Get
            Return _internalLocRepo
        End Get
    End Property

    ReadOnly Property Interactor() As IInteractor Implements IApplication.Interactor
        Get
            Return _interactor
        End Get
    End Property

    ''' <summary>
    ''' 
    ''' </summary>
    ''' <typeparam name="TMain"></typeparam>
    ''' <returns></returns>
    ''' <remarks>This builder is here because there is no such thing as generic constructors.</remarks>
    Public Shared Function Create(Of TMain As IContext)(
            Optional tokenParser As ITokenParser = Nothing,
            Optional interactor As IInteractor = Nothing,
            Optional intLocRepo As ILocalizationRepository = Nothing,
            Optional extLocRepo As ILocalizationRepository = Nothing
            ) As IApplication
        Dim app As New StandardApplication(GetType(TMain),
            tokenParser,
            interactor,
            intLocRepo,
            extLocRepo)
        Return app
    End Function

    ReadOnly Property CurrentContext As IContext
        Get
            Return CurrentContextNode.Value
        End Get
    End Property

    ReadOnly Property CurrentContextNode() As ITreeNode(Of IContext) Implements IApplication.CurrentContextNode
        Get
            Return _currentContextNode
        End Get
    End Property


    Private Sub SetRootContextNode(value As ITreeNode(Of IContext))
        If Not _rootContextNode Is value Then
            _rootContextNode = value
        End If
        _currentContextNode = value
    End Sub

    ReadOnly Property RootContextNode() As ITreeNode(Of IContext) Implements IApplication.RootContextNode
        Get
            Return _rootContextNode
        End Get
    End Property

    Public Function GetAllContexts() As IEnumerable(Of IContext) Implements IApplication.GetAllContexts
        If _contextsCache Is Nothing Then
            _contextsCache = CurrentContextNode.DepthPreorderTraverse()
        End If
        Return _contextsCache
    End Function

    Public Sub Run() Implements IApplication.Run
        Dim action = RootContextNode.Value.GetActions().GetNamedItem("credits")
        action.Execute()
        Do Until ExecuteCommand()
            DependencyResolver.ClearRequestScope()
        Loop
    End Sub

    Private Function ExecuteCommand() As Boolean Implements IApplication.ExecuteCommand
        ' Ask command line
        Dim command As String

        Dim prompt =
            {ApplicationPrompt}.Concat(
                CurrentContextNode.
                Ancestors.
                Reverse.
                Skip(1).
                Select(Function(node) node.Name)).
            JoinStr(" ")

        command = Interactor.AskCommand(prompt)

        ' Parse command line
        _tokenParser.Initialize()
        _tokenParser.Parse(command)

        If _tokenParser.Errors.Count > 0 Then
            Interactor.ShowErrors(_tokenParser.Errors.Select(
                                  Function(parseErr)
                                      Return InternalLocRepo.TransF("executecommand_parseerror", parseErr.Index, parseErr.StartIndex)
                                  End Function))
            Return False
        End If

        Dim saTokens = _tokenParser.Tokens

        If saTokens.Count = 0 Then
            Return False
        End If

        Dim i = 0

        ' Context retrieval (recursive)
        Dim executionCtxNode = CurrentContextNode
        Dim nextCtxNode As ITreeNode(Of IContext) = Nothing
        Do
            Dim contextName = saTokens(i)

            If contextName = ".." Then
                nextCtxNode = executionCtxNode.GetParent
                i += 1
            Else
                nextCtxNode = executionCtxNode.GetChildNodes.SingleOrDefault(Function(ctrl) ctrl.Value.IsNamed(contextName)).As(Of ITreeNode(Of IContext))()
            End If
            If nextCtxNode IsNot Nothing Then
                executionCtxNode = nextCtxNode
                i += 1
            End If
        Loop Until nextCtxNode Is Nothing OrElse i = saTokens.Count

        If i < saTokens.Count Then
            ' Action retrieval
            Dim actionName = saTokens(i).ToLower
            Dim action As IAction
            Dim ancestorExecutionCtx = executionCtxNode
            Do
                action = ancestorExecutionCtx.Value.GetActions().GetNamedItem(actionName)

                If action Is Nothing Then
                    ancestorExecutionCtx = ancestorExecutionCtx.GetParent
                    If ancestorExecutionCtx Is Nothing Then
                        Exit Do
                    End If
                Else
                    Exit Do
                End If
            Loop

            If action Is Nothing Then
                Interactor.ShowErrors({InternalLocRepo.TransF("executecommand_actionnotfound", actionName, executionCtxNode.Ancestors.Select(Function(ctx) ctx.Name).JoinStr(","))})
                Return False
            Else
                i += 1
                For j = i To saTokens.Count - 1
                    Interactor.TokenQueue.Enqueue(saTokens(j))
                Next
                action.Execute()
                Interactor.ShowMessage("")
                Return (action.Name = "quit")
            End If
        Else
            Interactor.ShowMessage(InternalLocRepo.TransF("use_message", executionCtxNode.Value.Name))
            _currentContextNode = executionCtxNode
            Return False
        End If
    End Function

End Class

File: Src/Icm.ContextConsole/Application/StandardApplication.cs
using System;
using System.Collections.Generic;
using System.Linq;
using Icm;
using Icm.Localization;
using Icm.Collections;
using Icm.Tree;

public class StandardApplication : IApplication
{
	private readonly ILocalizationRepository _internalLocRepo;
	private readonly ILocalizationRepository _externalLocRepo;
	private readonly IInteractor _interactor;
	private ITreeNode<IContext> _currentContextNode;
	private ITreeNode<IContext> _rootContextNode;
	private IEnumerable<IContext> _contextsCache;

	private readonly ITokenParser _tokenParser;
	public StandardApplication(Type rootContextType, ITokenParser tokenParser = null, IInteractor interactor = null, ILocalizationRepository intLocRepo = null, ILocalizationRepository extLocRepo = null) : this(new ReflectionContextTreeBuilder(rootContextType), tokenParser, interactor, intLocRepo, extLocRepo)
	{
	}

	public StandardApplication(TreeNode<Type> rootContextNode, ITokenParser tokenParser = null, IInteractor interactor = null, ILocalizationRepository intLocRepo = null, ILocalizationRepository extLocRepo = null) : this(new TypeContextTreeBuilder(rootContextNode), tokenParser, interactor, intLocRepo, extLocRepo)
	{
	}


	public StandardApplication(IContextTreeBuilder treeBuilder, ITokenParser tokenParser = null, IInteractor interactor = null, ILocalizationRepository intLocRepo = null, ILocalizationRepository extLocRepo = null)
	{
		if (tokenParser == null) {
			_tokenParser = new StandardTokenParser();
		} else {
			_tokenParser = tokenParser;
		}
		if (intLocRepo == null) {
			intLocRepo = new ResourceLocalizationRepository(My.Resources.Resources.ResourceManager);
		}
		if (extLocRepo == null) {
			extLocRepo = new DictionaryLocalizationRepository();
		}
		if (interactor == null) {
			interactor = new StreamsInteractor(intLocRepo);
		}
		if (tokenParser == null) {
			tokenParser = new StandardTokenParser();
		}
		treeBuilder.Application = this;
		_interactor = interactor;
		_internalLocRepo = intLocRepo;
		_externalLocRepo = extLocRepo;
		ApplicationPrompt = System.Reflection.Assembly.GetExecutingAssembly().CodeBase;
		DependencyResolver = new StandardDependencyResolver();
		// Stablish root context in the last place so that the IContext.Initialize routine
		// can access the former application values.
		SetRootContextNode(treeBuilder.GetTree());
	}

	public string ApplicationPrompt { get; set; }

	public IDependencyResolver DependencyResolver { get; set; }

	public ILocalizationRepository ExternalLocRepo {
		get { return _externalLocRepo; }
	}

	public ILocalizationRepository InternalLocRepo {
		get { return _internalLocRepo; }
	}

	public IInteractor Interactor {
		get { return _interactor; }
	}

	/// <summary>
	/// 
	/// </summary>
	/// <typeparam name="TMain"></typeparam>
	/// <returns></returns>
	/// <remarks>This builder is here because there is no such thing as generic constructors.</remarks>
	public static IApplication Create<TMain>(ITokenParser tokenParser = null, IInteractor interactor = null, ILocalizationRepository intLocRepo = null, ILocalizationRepository extLocRepo = null) where TMain : IContext
	{
		StandardApplication app = new StandardApplication(typeof(TMain), tokenParser, interactor, intLocRepo, extLocRepo);
		return app;
	}

	public IContext CurrentContext {
		get { return CurrentContextNode.Value; }
	}

	public ITreeNode<IContext> CurrentContextNode {
		get { return _currentContextNode; }
	}


	private void SetRootContextNode(ITreeNode<IContext> value)
	{
		if ((!object.ReferenceEquals(_rootContextNode, value))) {
			_rootContextNode = value;
		}
		_currentContextNode = value;
	}

	public ITreeNode<IContext> RootContextNode {
		get { return _rootContextNode; }
	}

	public IEnumerable<IContext> GetAllContexts()
	{
		if (_contextsCache == null) {
			_contextsCache = CurrentContextNode.DepthPreorderTraverse();
		}
		return _contextsCache;
	}

	public void Run()
	{
		var action = RootContextNode.Value.GetActions().GetNamedItem("credits");
		action.Execute();
		while (!ExecuteCommand()) {
			DependencyResolver.ClearRequestScope();
		}
	}

	public bool ExecuteCommand()
	{
		// Ask command line
		string command = null;

		var prompt = new[] { ApplicationPrompt }
        .Concat(CurrentContextNode.Ancestors()
        .Reverse()
        .Skip(1)
        .Select(node => node.Name()))
        .JoinStr(" ");

		command = Interactor.AskCommand(prompt);

		// Parse command line
		_tokenParser.Initialize();
		_tokenParser.Parse(command);

		if (_tokenParser.Errors.Any()) {
			Interactor.ShowErrors(_tokenParser.Errors.Select(parseErr => InternalLocRepo.TransF("executecommand_parseerror", parseErr.Index, parseErr.StartIndex)));
			return false;
		}

		var saTokens = _tokenParser.Tokens.ToArray();

		if (!saTokens.Any()) {
			return false;
		}

		var i = 0;

		// Context retrieval (recursive)
		var executionCtxNode = CurrentContextNode;
		ITreeNode<IContext> nextCtxNode = null;
		do {
			var contextName = saTokens[i];

			if (contextName == "..") {
				nextCtxNode = executionCtxNode.GetParent();
				i += 1;
			} else {
				nextCtxNode = executionCtxNode.GetChildNodes().SingleOrDefault(ctrl => ctrl.Value.IsNamed(contextName)).As<ITreeNode<IContext>>();
			}
			if (nextCtxNode != null) {
				executionCtxNode = nextCtxNode;
				i += 1;
			}
		} while (!(nextCtxNode == null || i == saTokens.Count()));

		if (i < saTokens.Count()) {
			// Action retrieval
			var actionName = saTokens[i].ToLower();
			IAction action = default(IAction);
			var ancestorExecutionCtx = executionCtxNode;
			do {
				action = ancestorExecutionCtx.Value.GetActions().GetNamedItem(actionName);

				if (action == null) {
					ancestorExecutionCtx = ancestorExecutionCtx.GetParent();
					if (ancestorExecutionCtx == null) {
						break; // TODO: might not be correct. Was : Exit Do
					}
				} else {
					break; // TODO: might not be correct. Was : Exit Do
				}
			} while (true);

			if (action == null) {
				Interactor.ShowErrors(new [] { InternalLocRepo.TransF("executecommand_actionnotfound", actionName, executionCtxNode.Ancestors().Select(ctx => ctx.Name()).JoinStr(",")) });
				return false;
			} else {
				i += 1;
				for (var j = i; j <= saTokens.Count() - 1; j++) {
					Interactor.TokenQueue.Enqueue(saTokens[j]);
				}
				action.Execute();
				Interactor.ShowMessage("");
				return (action.Name() == "quit");
			}
		} else {
			Interactor.ShowMessage(InternalLocRepo.TransF("use_message", executionCtxNode.Value.Name()));
			_currentContextNode = executionCtxNode;
			return false;
		}
	}

}

//=======================================================
//Service provided by Telerik (www.telerik.com)
//Conversion powered by NRefactory.
//Twitter: @telerik
//Facebook: facebook.com/telerik
//=======================================================
